src/charm.py

```diff
@@ -675,6 +845,38 @@ def _routing_mode(self) -> _RoutingMode:
         """
         return _RoutingMode(self.config["routing_mode"])
 
+    @property
+    def version(self) -> Optional[str]:
+        """Return the workload version."""
+        if not self.container.can_connect():
+            return None
+
+        version_output, _ = self.container.exec([self._bin_path, "version"]).wait_output()
+        # Output looks like this:
+        # Version:      2.9.6
+        # Codename:     banon
+        # Go version:   go1.18.9
+        # Built:        2022-12-07_04:28:37PM
+        # OS/Arch:      linux/amd64
+
+        if result := re.search(r"Version:\s*(\d*\.\d*\.\d*)", version_output):
+            return result.group(1)
+        return None
+
+    def _set_workload_version(self):
+        if version := self.version:
+            self.unit.set_workload_version(version)
+        else:
+            logger.debug(
+                "Cannot set workload version at this time: could not get Traefik version."
+            )
+
+    @property
+    def cert_subject(self) -> str:
+        """Provide certificate subject."""
+        # TODO: Use a good default for when the external_hostname is None
```

Comment:
A good default is probably "don't enable TLS if the prerequisite configuration isn't present." Having a subject with a cluster-local DNS name is gonna be a bad time for anyone who deploys Traefik, doesn't configure this, and their browser upgrades a connection to HTTPS only to immediately see an invalid cert.

Reply:
Modified the logic here a bit. Wdyt?

Reply:
:+1: Testing the other path (if a `LoadBalancer` has a hostname set) will almost need to be a little ugly in integration tests (it's mostly "cloud" load balancers which set this), but that's probably ok.